Fixed #28546 -- Fixed translation's to_locale() with langauge subtags.

bwh91 opened the pull request:
Added statement to check for subtags and return correct formatting. The statement to get correct locale with 'sr-latn' did not allow the subtags to be formatted properly. I added an if statement to catch the subtags.
I also added to the 'to_locale' test to check for proper subtag formatting.
https://code.djangoproject.com/ticket/28546

claudep commented:
Looks like your branch needs a rebase!

bwh91 commented:
Sorry about that. Its rebased now.

timgraham commented on tests/i18n/tests.py:
```diff
@@ -259,8 +259,14 @@ def test_maclines(self):
             self.assertEqual('Catalan Win\nEOF\n', gettext('Win\r\nEOF\r\n'))
 
     def test_to_locale(self):
+        """
```
I'd use inline comments rather than a docstring. I also added a failing case:
```
tests = (
    ('en-us', 'en_US'),
    # A langage with > 2 characters after the dash.
    ('sr-lat', 'sr_Lat'),
    # Languages with a private-use subtag (x-informal).
    ('nl-nl-x-informal', 'nl_NL-x-informal'),
    ('sr-lat-x-informal', 'sr_LAT-x-informal'),
)
for lang, output in tests:
    with self.subTest(lang=lang):
        self.assertEqual(to_locale(lang), output)
```

timgraham commented on django/utils/translation/trans_real.py:
```diff
@@ -67,8 +67,11 @@ def to_locale(language, to_lower=False):
         if to_lower:
             return language[:p].lower() + '_' + language[p + 1:].lower()
         else:
+            # checking for subtags
```
`# Check for subtags, e.g. 'ab-cd-x-informal'.`

bwh91 commented:
I have added your suggestions as well as corrected the incorrect formatting when the language code is in this format 'sr-latn-x-informal'.

timgraham commented on tests/i18n/tests.py:
```diff
@@ -259,8 +259,21 @@ def test_maclines(self):
-        self.assertEqual(to_locale('en-us'), 'en_US')
-        self.assertEqual(to_locale('sr-lat'), 'sr_Lat')
+        """
+        Tests the to_locale function and the special case of Serbian Latin as well as subtags
```
The inline comments are sufficient. I don't think the docstring is adding additional value.

bwh91 commented:
Ok. Cleaned up the function. I think it is a lot easier to read. I also added another test as I realized that when Django first starts it uses the to_local function with a call like "to_locale('en')" It broke on me and there wasn't a test for it.

timgraham commented:
I added some more test cases and simplifications in https://github.com/django/django/pull/9040. I'd appreciate your review, thanks!
